skrf/network.py

```diff
@@ -4381,6 +4381,120 @@ def stability_circle(self, target_port: str, npoints: int = 181) -> npy.ndarray:
         sc = sc_real + 1j * sc_imag
         return sc.T
 
+    def gain_circle(self, target_port: str, gain: float, npoints: int = 181) -> npy.ndarray:
+        r"""
+        Returns gain circles for a given port ('source' or 'load') and a specified gain.
+        The center and radius of the source gain circle are calculated by the following equations.
+
+        .. math::
+
+                C_{S} = \frac{g_{S}S_{11}^*}{1 - (1 - g_{S})|S_{11}|^{2}}
+
+                R_{S} = |\frac{\sqrt{(1 - g_{S})}(1 - |S_{11}|^{2})}{1 - (1 - g_{S})|S_{11}|^{2}}
+
+        where :math:`g_{S}` is obtained by normalizing the specified gain by the maximum gain of the source matching network :math:`G_{Smax}`
+
+        .. math::
+
+                g_{S} = \frac{gain}{G_{Smax}} = gain * (1 - |S_{11}|^{2})
+
+        Similarly, those of the load side are calculated by the following equations.
+
+        .. math::
+
+                C_{L} = \frac{g_{L}S_{22}^*}{1 - (1 - g_{L})|S_{22}|^{2}}
+
+                R_{L} = |\frac{\sqrt{(1 - g_{L})}(1 - |S_{22}|^{2})}{1 - (1 - g_{L})|S_{22}|^{2}}
+
+                with
+
+                g_{L} = \frac{gain}{G_{Lmax}} = gain * (1 - |S_{22}|^{2})
+
+        Parameters
+        ----------
+        target_port : str
+            Specifies the 'source' or 'load' side to calculate gain circles.
+        gain : float
+            Gain of source or load matching network in decibels.
+        npoints : int, optional
+            The number of points on the circumference of the circle.
+            More points result in a smoother circle, but require more computation. Default is 181.
+
+        Returns
+        -------
+        ntwk : :class:`numpy.ndarray` (shape is `npoints x f`)
+            Gain circle in complex numbers
+
+        Example
+        --------
+        >>> import skrf as rf
+        >>> import matplotlib.pyplot as plt
+
+        Create a network object
+
+        >>> ntwk = rf.Network('fet.s2p')
+
+        Calculate the source gain circles for all the frequencies
+
+        >>> sgc = ntwk.gain_circle(target_port='source', gain=2.0)
+
+        Plot the circles on the smith chart
+
+        >>> rf.plotting.plot_smith(s=sgc, smith_r=1)
+        >>> plt.show()
+
+        Slicing the network allows you to specify a frequency
+
+        >>> sgc = ntwk['1GHz'].gain_circle(target_port='source', gain=2.0)
+        >>> rf.plotting.plot_smith(s=sgc, smith_r=1)
+        >>> plt.show()
+
+        References
```

Comment:
Providing references is very good. But they are not referenced in any of the docstring sections above, are they? (Not sure if they have to be referenced...)

Reply:
I added reference numbers in the docstring section.